File: New-EMS-de9b154a3f16b49898b382f5d7bad10bd9734b51/server/check_braces.py

import sys

def check_brace_balance(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    stack = []
    line_num = 1
    col_num = 0
    
    in_single_quote = False
    in_double_quote = False
    in_template_literal = False
    in_line_comment = False
    in_block_comment = False
    
    for i, char in enumerate(content):
        if char == '\n':
            line_num += 1
            col_num = 0
            in_line_comment = False
            continue
        
        col_num += 1
        
        if in_line_comment:
            continue
            
        if in_block_comment:
            if char == '/' and content[i-1] == '*':
                in_block_comment = False
            continue
            
        if in_single_quote:
            if char == "'" and content[i-1] != '\\':
                in_single_quote = False
            continue
            
        if in_double_quote:
            if char == '"' and content[i-1] != '\\':
                in_double_quote = False
            continue
            
        if in_template_literal:
            if char == '`' and content[i-1] != '\\':
                in_template_literal = False
            continue
            
        # Check for comments start
        if char == '/' and i + 1 < len(content):
            if content[i+1] == '/':
                in_line_comment = True
                continue
            if content[i+1] == '*':
                in_block_comment = True
                continue
                
        # Check for quotes start
        if char == "'":
            in_single_quote = True
            continue
        if char == '"':
            in_double_quote = True
            continue
        if char == '`':
            in_template_literal = True
            continue
            
        if char == '{':
            stack.append(('{', line_num, col_num))
        elif char == '}':
            if not stack:
                print(f"Extra closing brace '}}' at line {line_num}, col {col_num}")
                # Print context
                start_line = max(1, line_num - 5)
                end_line = line_num + 5
                print(f"Context (lines {start_line}-{end_line}):")
                lines = content.split('\n')
                for j in range(start_line-1, min(len(lines), end_line)):
                    print(f"{j+1}: {lines[j]}")
                return
            stack.pop()
        elif char == '(':
            stack.append(('(', line_num, col_num))
        elif char == ')':
            if not stack or stack[-1][0] != '(':
                # We only track braces strictly but sometimes mismatch in parens causes issues too
                pass
            else:
                stack.pop()

    if stack:
        for item in stack:
            print(f"Unclosed '{item[0]}' starting at line {item[1]}, col {item[2]}")
    else:
        print("Braces are balanced (strictly speaking)")

if __name__ == "__main__":
    check_brace_balance(sys.argv[1])

File: New-EMS-de9b154a3f16b49898b382f5d7bad10bd9734b51/server/check_braces.js

const fs = require('fs');

const file_path = process.argv[2];
const content = fs.readFileSync(file_path, 'utf8');

let stack = [];
let line_num = 1;
let col_num = 0;

let in_single_quote = false;
let in_double_quote = false;
let in_template_literal = false;
let in_line_comment = false;
let in_block_comment = false;

for (let i = 0; i < content.length; i++) {
    const char = content[i];
    if (char === '\n') {
        line_num++;
        col_num = 0;
        in_line_comment = false;
        continue;
    }
    col_num++;

    if (in_line_comment) continue;

    if (in_block_comment) {
        if (char === '/' && content[i - 1] === '*') {
            in_block_comment = false;
        }
        continue;
    }

    if (in_single_quote) {
        if (char === "'" && content[i - 1] !== '\\') {
            in_single_quote = false;
        }
        continue;
    }

    if (in_double_quote) {
        if (char === '"' && content[i - 1] !== '\\') {
            in_double_quote = false;
        }
        continue;
    }

    if (in_template_literal) {
        if (char === '`' && content[i - 1] !== '\\') {
            in_template_literal = false;
        }
        continue;
    }

    // Check for comments start
    if (char === '/' && i + 1 < content.length) {
        if (content[i + 1] === '/') {
            in_line_comment = true;
            continue;
        }
        if (content[i + 1] === '*') {
            in_block_comment = true;
            continue;
        }
    }

    // Check for quotes start
    if (char === "'") {
        in_single_quote = true;
        continue;
    }
    if (char === '"') {
        in_double_quote = true;
        continue;
    }
    if (char === '`') {
        in_template_literal = true;
        continue;
    }

    if (char === '{') {
        stack.push({ char: '{', line: line_num, col: col_num });
    } else if (char === '}') {
        if (stack.length === 0) {
            console.log(`Extra closing brace '}' at line ${line_num}, col ${col_num}`);
            printContext(content, line_num);
            process.exit(0);
        }
        stack.pop();
    }
}

if (stack.length > 0) {
    stack.forEach(item => {
        console.log(`Unclosed '${item.char}' starting at line ${item.line}, col ${item.col}`);
    });
} else {
    console.log("Braces are balanced (strictly speaking)");
}

function printContext(text, line) {
    const lines = text.split('\n');
    const start = Math.max(0, line - 6);
    const end = Math.min(lines.length, line + 5);
    for (let i = start; i < end; i++) {
        console.log(`${i + 1}: ${lines[i]}`);
    }
}
